src/main/kotlin/com/zepben/evolve/services/network/EquivalentNetworkUtils.kt

```diff
@@ -0,0 +1,465 @@
+            else -> ce.baseVoltage
+        }
+
+    private fun addEquivalentNetwork(
+        network: NetworkService,
+        edgeNode: ConnectivityNode,
+        createEquivalentBranches: (EquivalentBranchDetails) -> Sequence<Pair<PhaseCode, EquivalentBranch>>,
+        createEquivalentEquipment: (EquivalentEquipmentDetails) -> Sequence<Pair<PhaseCode, ConductingEquipment>>
+    ): EquivalentNetworkConnection {
+        val edgeEquipment = edgeNode.terminals.first().conductingEquipment
+            ?: throw IllegalStateException("ConnectivityNode for EquivalentBranch must have ConductingEquipment")
+        val containersForEquivalentNetwork = getContainersForEdgeConnectivityNode(edgeNode)
+        val branchToEquipment = createEquivalentBranches(edgeEquipment via edgeNode).map { (phaseCode, equivalentBranch) ->
+            equivalentBranch.apply {
+                network.add(this)
+                containersForEquivalentNetwork.forEach {
+                    it.addEquipment(this)
+                    addContainer(it)
+                }
+
+                Terminal("${this.mRID}-t${this.numTerminals() + 1}").also {
+                    network.add(it)
+                    addTerminal(it)
+                    it.conductingEquipment = this
+                    it.phases = phaseCode
+                }
+                network.connect(this.terminals.first(), edgeNode.mRID)
+            } to
+                createEquivalentEquipment(edgeEquipment via edgeNode via equivalentBranch).map { (phaseCode, conEq) ->
+                    conEq.apply {
+                        network.tryAdd(this)
+                        containersForEquivalentNetwork.forEach {
+                            it.addEquipment(this)
+                            this.addContainer(it)
+                        }
+
+                        val eqBranchTerminal = Terminal("${equivalentBranch.mRID}-t${equivalentBranch.numTerminals() + 1}").also {
+                            network.add(it)
+                            equivalentBranch.addTerminal(it)
+                            it.conductingEquipment = equivalentBranch
+                            it.phases = phaseCode
+                        }
+                        val conEqTerminal = Terminal("${this.mRID}-t${this.numTerminals() + 1}").also {
+                            network.add(it)
+                            conEq.addTerminal(it)
+                            it.conductingEquipment = this
+                            it.phases = phaseCode
+                        }
+                        network.connect(eqBranchTerminal, conEqTerminal)
+                    }
+                }.toSet()
+        }.toMap()
+
+        return EquivalentNetworkConnection(edgeEquipment, edgeNode, branchToEquipment)
+    }
+
+    private fun getContainersForEdgeConnectivityNode(cn: ConnectivityNode): Set<EquipmentContainer> =
+        ((cn.terminals as List).first().conductingEquipment?.containers?.toSet() ?: emptySet()).minus(
```

Comment:
why are the `as List`s necessary?  Tests seem to pass without them.

Reply:
This was probably intellij being smart about something is my guess and I just missed it.